### PR TITLE
generalized liquid pooler

### PR DESCRIPTION
**Pull request by bekauz**
attempt at #118 

Instead of using the notion of native/liquid staked tokens, this liquid pooler deals with denom `a` and denom `b`.

because of no instantiate2 availability, after instantiation this contract expects a migration config update that would set the holder address. it's necessary for directing the LP tokens when providing liquidity to the pool. this is done by the covenant itself.

merging into the two-party-pol branch as it is blocked by this. I expect changes in design on this one so no unit tests for now. Only happy path is included in the interchaintests.

### CODE REVIEW

File: contracts/astroport-liquid-pooler/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+        .query_all_balances(env.contract.address.to_string())?;
+    let (coin_a, coin_b) = get_relevant_balances(
+        bal_coins,
+        asset_data.asset_a_denom.as_str(),
+        asset_data.asset_b_denom.as_str(),
+    );
+
+    // depending on available balances we attempt a different action:
+    match (coin_a.amount.is_zero(), coin_b.amount.is_zero()) {
+        // one balance is non-zero, we attempt single-side
```
Comment (uditvira) [inline contracts/astroport-liquid-pooler/src/contract.rs:142]:
```suggestion
        // exactly one balance is non-zero, we attempt single-side
```

File: contracts/astroport-liquid-pooler/src/msg.rs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+#[cw_serde]
+pub struct InstantiateMsg {
+    pub pool_address: String,
+    pub clock_address: String,
+    pub slippage_tolerance: Option<Decimal>,
+    pub autostake: Option<bool>,
+    pub assets: AssetData,
+    pub single_side_lp_limits: SingleSideLpLimits,
+    pub expected_pool_ratio: Decimal,
+    pub acceptable_pool_ratio_delta: Decimal,
```
[the remark this answers is not in the document]
Comment (uditvira) [reply contracts/astroport-liquid-pooler/src/msg.rs:17]: We can hold on this for now, the way you have it currently works. What I'm suggesting above may have to be calculated differently based on pool types. Prices follow different curves and you'll need more info about the pool

**Review by uditvira — APPROVED** [review]
Nice work. Easy to follow code.

File: contracts/astroport-liquid-pooler/src/msg.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
+use astroport::asset::{Asset, AssetInfo};
+use cosmwasm_schema::{cw_serde, QueryResponses};
+use cosmwasm_std::{Addr, Attribute, Binary, Decimal, Uint128};
+use covenant_macros::{clocked, covenant_clock_address, covenant_deposit_address};
+
+use crate::error::ContractError;
+
+#[cw_serde]
+pub struct InstantiateMsg {
```
[the remark this answers is not in the document]
Comment (bekauz) [reply contracts/astroport-liquid-pooler/src/msg.rs:9]: Solid. Will add this.

File: contracts/astroport-liquid-pooler/src/msg.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bekauz) [reply contracts/astroport-liquid-pooler/src/msg.rs:17]: So, `slippage_tolerance` in the astroport `ProvideLiquidity` message is a `Decimal`, same as `acceptable_pool_ratio_delta`. I suppose we could do something like this, although I'm not 100% sure if that's the right move @uditvira :
1. find current pool ratio
2. find the deviation of the current pool ratio from our `expected_pool_ratio`
  2.a deviation > `acceptable_pool_ratio_delta` -> error
  2.b deviation <= `acceptable_pool_ratio_delta` -> we find the (abs) diff between acceptable and current delta and set that to be our slippage tolerance..?

File: contracts/astroport-liquid-pooler/examples/schema.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+fn main() {
+    let mut out_dir = current_dir().unwrap();
+    out_dir.push("schema");
+    create_dir_all(&out_dir).unwrap();
+    remove_schemas(&out_dir).unwrap();
+
+    export_schema(&schema_for!(InstantiateMsg), &out_dir);
+    export_schema(&schema_for!(QueryMsg), &out_dir);
+    export_schema(&schema_for!(ExecuteMsg), &out_dir);
+}
```
Comment (Art3miX) [inline contracts/astroport-liquid-pooler/examples/schema.rs:15]: Theres new way to do schema.

File: contracts/astroport-liquid-pooler/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+    let asset_data = ASSETS.load(deps.storage)?;
+    let lp_config = LP_CONFIG.load(deps.storage)?;
+
+    // validate that the pool did not migrate to a new pair type
+    let pool_response: PairInfo = deps
+        .querier
+        .query_wasm_smart(&lp_config.pool_address, &astroport::pair::QueryMsg::Pair {})?;
+    if pool_response.pair_type != lp_config.pair_type {
+        return Err(ContractError::PairTypeMismatch {})   
+    }
```
Comment (Art3miX) [inline contracts/astroport-liquid-pooler/src/contract.rs:125]: interested to why this matters?

File: contracts/astroport-liquid-pooler/src/contract.rs [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+                msg: to_binary(&single_sided_liq_msg)?,
+                funds: vec![coin_a],
+            }),
+            SINGLE_SIDED_REPLY_ID,
+        );
+
+        return Ok(Some(submsg));
+    }
+
+    // if neither a nor b token single side lp message was built, we just go back and wait
```
Comment (Art3miX) [inline contracts/astroport-liquid-pooler/src/contract.rs:330]: What if the amount is higher then `single_side_lp_limit`? in this code the funds will be stuck until we unstuck them using a migration

**Review by Art3miX — COMMENTED** [review]
I love how neat your code is.

**Pull request merged**